### PR TITLE
Pathing

### PR DESCRIPTION
**Pull request by DrPeeper**

### CODE REVIEW

File: lib/nav.py [hunk 1]
```diff
@@ -36,3 +105,90 @@ def force_dir(dir, n=1):
...
+			self.force_dir(DIRS.FORWARD, to_go * -1)
+
+		# move z
+		to_go = location[2] - self.location[2]
+		if to_go > 0:
+			self.force_dir(DIRS.UP, to_go)
+		elif to_go < 0:
+			self.force_dir(DIRS.DOWN, to_go * -1)
+
+		return True if self.location == location else False
```
Comment (maxtrussell) [inline lib/nav.py]: `return self.location == location`

File: lib/nav.py [hunk 2]
```diff
@@ -10,7 +10,72 @@ class DIRS(Enum):
...
+	},
+	CARDINALS.SOUTH: {
+		TURNS.LEFT:CARDINALS.EAST,
+		TURNS.RIGHT:CARDINALS.WEST,
+	},
+	CARDINALS.WEST: {
+		TURNS.LEFT:CARDINALS.SOUTH,
+		TURNS.RIGHT:CARDINALS.NORTH,
+	},
+}
```
Comment (maxtrussell) [inline lib/nav.py]: Couldn't you just use assign north east south and west to 0 1 2 and 3 respectively and do modular arithmetic?

turning right is 1 and left is -1

That's a standard way to handle navigation with cardinal directions without hardcoding

For example, turtle is facing east (val of 1), he wants to face south (val of 2). 

south - east => (2 - 1) => 1, so it needs one right turn

Another example, turtle is facing west, wants to face north
north - west = 0 - 3 = -3, but for anything with an absolute value of > 2, we can just do one of the opposite turn instead. So 1 right turn

The logic ends up looking like this:
```
def get_turns(curr, target) -> int:
  """ A positive result of n, means n right turns, a negative n means n left turns """
  right_turns = (target-curr) % 4
  left_turns = (curr-target) % 4
  return right_turns if right_turns < left_turns else -left_turns
```

File: bin/navigate.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+turtle.dir(nav.DIRS.FORWARD, 5)
+test('Expected State: [0,6,0], NORTH')
+
+turtle.turn_to(nav.CARDINALS.EAST)
+test('Expected State: [0,6,0], EAST')
+
+turtle.dir(nav.DIRS.FORWARD, 5)
+test('Expected State: [5,6,0], EAST')
+
+print(turtle.go_to([0,0,0]))
```
Comment (maxtrussell) [inline bin/navigate.py]: You mind putting this file under a new directory 'tests' or something?

I don't want to clutter bin with tests

File: bin/schematic.py [hunk 4]
```diff
@@ -39,12 +40,13 @@ def parse_schematic(path, open_func):
 
 def build_slice(slice, block_defs):
     selected = 0
-    turn_dir = turtle.turnRight
+    turn_dir = nav.TURNS.RIGHT
+    turn = navigator.turn
```
Comment (maxtrussell) [inline bin/schematic.py]: No need to assign like this. Just call `navigator.turn` later

File: bin/schematic.py [hunk 5]
```diff
@@ -107,3 +109,5 @@ def count_col(col):
     # obviously you can only build from within minecraft
     for slice in schematic:
         build_slice(slice, block_defs)
+    navigator.go_to([0,0,0])
```
Comment (maxtrussell) [inline bin/schematic.py]: Your navigator knows how to get back to 0,0,0 but it has no guarantee that its path won't be obstructed.

This line looks like the path would just cut straight through the building that was just created.

File: lib/dig.py [hunk 6]
```diff
@@ -12,46 +12,51 @@
         'minecraft:lapis_lazuli':True,
 }
 
+navigator = nav.Navigator()
```
Comment (maxtrussell) [inline lib/dig.py]: Would we want this namespaced to quarry?

With this being global, I can imagine some very confusing cases where its state is modified by other functions

**Review by maxtrussell — CHANGES_REQUESTED** [review]

File: bin/schematic.py (hunk 4 above)
Comment (maxtrussell) [reply bin/schematic.py:44]: Also is there any reason navigator needs to handle basic left and right turns?

Seems like turtle.turnLeft and turtle.turnRight do the exact same thing

File: lib/nav.py (hunk 2 above)
Comment (DrPeeper) [reply lib/nav.py:78]: Will do thank you

**Pull request closed without merging**